qiskit/quantum_info/operators/dihedral/dihedral.py

```diff
@@ -0,0 +1,495 @@
+        if validate and not self.is_cnotdihedral():
+            raise QiskitError('Invalid CNOTDihedsral element.')
+
+    def _z2matmul(self, left, right):
+        """Compute product of two n x n z2 matrices."""
+        prod = np.mod(np.dot(left, right), 2)
+        return prod
+
+    def _z2matvecmul(self, mat, vec):
+        """Compute mat*vec of n x n z2 matrix and vector."""
+        prod = np.mod(np.dot(mat, vec), 2)
+        return prod
+
+    def __mul__(self, other):
+        """Left multiplication self * other."""
+        if self.num_qubits != other.num_qubits:
+            raise QiskitError("Multiplication on different number of qubits.")
+        result = CNOTDihedral(self.num_qubits)
+        result.shift = [(x[0] + x[1]) % 2
+                        for x in zip(self._z2matvecmul(self.linear, other.shift), self.shift)]
+        result.linear = self._z2matmul(self.linear, other.linear)
+        # Compute x' = B1*x + c1 using the p_j identity
+        new_vars = []
+        for i in range(self.num_qubits):
+            support = np.arange(self.num_qubits)[np.nonzero(other.linear[i])]
+            poly = SpecialPolynomial(self.num_qubits)
+            poly.set_pj(support)
+            if other.shift[i] == 1:
+                poly = -1 * poly
+                poly.weight_0 = (poly.weight_0 + 1) % 8
+            new_vars.append(poly)
+        # p' = p1 + p2(x')
+        result.poly = other.poly + self.poly.evaluate(new_vars)
+        return result
+
+    def __rmul__(self, other):
+        """Right multiplication other * self."""
+        if self.num_qubits != other.num_qubits:
+            raise QiskitError("Multiplication on different number of qubits.")
+        result = CNOTDihedral(self.num_qubits)
+        result.shift = [(x[0] + x[1]) % 2
+                        for x in zip(self._z2matvecmul(other.linear,
+                                                       self.shift),
+                                     other.shift)]
+        result.linear = self._z2matmul(other.linear, self.linear)
+        # Compute x' = B1*x + c1 using the p_j identity
+        new_vars = []
+        for i in range(self.num_qubits):
+            support = np.arange(self.num_qubits)[np.nonzero(self.linear[i])]
+            poly = SpecialPolynomial(self.num_qubits)
+            poly.set_pj(support)
+            if other.shift[i] == 1:
+                poly = -1 * poly
+                poly.weight_0 = (poly.weight_0 + 1) % 8
+            new_vars.append(poly)
+        # p' = p1 + p2(x')
+        result.poly = self.poly + other.poly.evaluate(new_vars)
+        return result
+
+    @property
+    def key(self):
+        """Return a string representation of a CNOT-dihedral object."""
+        tup = (self.poly.key, tuple(map(tuple, self.linear)),
+               tuple(self.shift))
+        return str(tup)
+
+    def __eq__(self, x):
+        """Test equality."""
+        return isinstance(x, CNOTDihedral) and self.key == x.key
```

Comment:
Rather than call `key` this should either have the functionality of that method directly here, or do a check of each component `self.poly == other.poly and self.linear == other.linear and self.shift == other.shift`